feat: Implement Model Reasoning/Thinking Stream

johannes117 opened the pull request:
This PR introduces support for streaming and displaying the reasoning or "thinking" process from AI models that support this feature (e.g., Gemini 2.5, Claude 4). This allows users to see the model's thought process in real-time before it generates the final answer.

#### Key Changes:

*   **Backend Support:**
    *   The Convex backend now processes reasoning streams from providers like Google and Anthropic.
    *   A new `reasoning` message part type has been added to store and render this information.
    *   Model configurations have been updated with `supportsReasoning` and `canToggleThinking` flags.

*   **"Thinking" UI Toggle:**
    *   A new "Thinking" section with a toggle switch has been added to the "Run Settings" panel.
    *   This allows users to enable or disable reasoning for models where it's optional. For models where it's always on, the UI reflects this.

*   **Reasoning Display:**
    *   A new collapsible `MessageReasoning` component has been added to the chat interface. It displays the "Reasoning" steps and includes a loading indicator while streaming.

*   **Improved Loading Indicator:**
    *   The loading indicator logic has been refined to only appear before *any* content (including text, tool calls, or reasoning) has been streamed back from the assistant.

johannes117 committed:
feat: Implement attachment management and multi-modal message handling

- Added support for file attachments, including upload, retrieval, and deletion functionalities.
- Introduced new mutations and queries for managing attachments in the Convex backend.
- Enhanced message processing to handle multi-modal content, including text and images.
- Updated ChatInput component to allow file uploads and display previews.
- Refactored message validation to include image types and attachment references.
- Improved UI components to manage and display attachments effectively.

johannes117 committed:
feat: Enhance attachment management and conversation handling

- Added newAttachmentIds to ChatParams for tracking attachments in messages.
- Implemented internal mutation to update token counts for attachments.
- Introduced query to retrieve attachments for specific conversations.
- Updated message handling to link attachments to conversations and track their token usage.
- Enhanced UI components to display attachments and their token counts in the chat settings.

johannes117 committed:
feat: Update ChatRunSettings to include messages prop for enhanced attachment handling

- Modified ChatRunSettings component to accept messages as a prop, allowing for conditional fetching of attachments based on message presence.
- Improved rendering logic to display appropriate states for new chats and loading attachments, enhancing user experience in the chat interface.

johannes117 committed:
feat: Refactor Settings and Attachments management

- Introduced a new Tabs component in the Settings page to organize account, API keys, and attachments sections.
- Removed the AttachmentsPage route and integrated its functionality directly into the Settings component.
- Updated routing to eliminate the standalone Attachments page, enhancing user navigation.
- Added attachment handling features, including upload, download, and deletion capabilities within the Settings interface.
- Updated package.json to include the latest version of "@tavily/core" and adjusted dependencies accordingly.

johannes117 committed:
style: Refine ChatInput and ChatRunSettings components for improved UI consistency

- Removed shadow from ChatInput component for a cleaner look.
- Adjusted ChatRunSettings component by removing unnecessary shadow, enhancing visual clarity.
- Updated styles to ensure better alignment and responsiveness across different screen sizes.

johannes117 committed:
style: Update ChatInput and token limits for improved UI and functionality

- Enhanced ChatInput component styles for better visual consistency, including updated colors and hover effects.
- Adjusted placeholder text for clarity in user messaging.
- Commented out the "o3" model in token limits to reflect verification requirements, ensuring alignment with organizational policies.

johannes117 committed:
style: Update ChatInput component for improved visual consistency

- Enhanced the styling of the ChatInput component by updating background colors, borders, and shadows for a more cohesive design.
- Improved responsiveness and visual feedback during drag-and-drop interactions, ensuring a better user experience.

johannes117 committed:
style: Update ChatSidebar component for improved visual consistency

- Adjusted the SidebarHeader and SidebarFooter components by removing border styles for a cleaner look.
- Enhanced overall styling to align with recent design updates, ensuring a more cohesive user experience.

johannes117 committed:
feat: Implement branching functionality for conversations

- Added a new mutation to create branched conversations from existing ones, allowing users to maintain conversation context while exploring different discussion paths.
- Updated the conversation schema to include fields for branching status and origin details.
- Enhanced the ChatSidebar and MessageControls components to support visual indicators and actions for branched conversations.
- Included a new tooltip to inform users about the origin of branched conversations.
- Updated package.json to include the latest version of "@types/uuid" for UUID generation.

johannes117 committed:
fix: Update lastMessageAt timestamp in conversation branching logic

- Changed the lastMessageAt field to use the current timestamp instead of the last message's createdAt value, ensuring accurate tracking of the last message time during conversation branching.

johannes117 committed:
feat: Add reasoning message type and update related components

- Introduced a new message part type 'reasoning' to enhance message structure, allowing for reasoning content to be included in conversations.
- Updated message validators and types to accommodate the new reasoning part.
- Enhanced the Message and MessageReasoning components to render reasoning content appropriately, improving user interaction and clarity in conversations.
- Adjusted styles for better presentation of reasoning messages.

johannes117 committed:
feat: Enhance reasoning capabilities in chat interactions

- Added provider-specific options for reasoning in the AI model, enabling enhanced thought processes for Google and Anthropic models.
- Updated the Chat component to handle and display reasoning parts appropriately, improving user engagement with reasoning content.
- Removed outdated reasoning rendering logic from the Message component to streamline message handling.

johannes117 committed:
style: Simplify reasoning label in MessageReasoning component

- Changed the label from "💭 Reasoning" to "Reasoning" for a cleaner presentation.
- Maintained character count display for clarity in message context.

johannes117 committed:
feat: Update AI model configuration to support reasoning

- Added a `supportsReasoning` property to the ModelConfig type to indicate which models can utilize reasoning features.
- Updated specific AI models (Gemini 2.5 Pro, Gemini 2.5 Flash, Claude 4 Sonnet, Claude 4 Opus, o4-mini, DeepSeek R1) to enable reasoning capabilities.
- Refactored reasoning logic in the chat action to utilize the new configuration, streamlining provider-specific options for models that support reasoning.

johannes117 committed:
refactor: Remove character count display from MessageReasoning component

- Eliminated the character count display in the MessageReasoning component for a cleaner interface.
- Maintained the overall functionality and presentation of reasoning content.

johannes117 committed:
feat: Enhance reasoning handling in chat interactions

- Updated the chat action to pass reasoning content directly, improving the handling of reasoning updates.
- Modified the message schema to include an optional reasoning field for better tracking of reasoning text.
- Adjusted the Message and MessageReasoning components to render reasoning content more effectively, including support for streaming reasoning.
- Improved loading indicators in the Messages component to reflect the assistant's generating state accurately.

johannes117 committed:
refactor: Update reasoning handling in Message components

- Removed console log from the chat action to clean up output.
- Enhanced the Message and MessageReasoning components to better manage reasoning streaming states.
- Adjusted variable names for clarity in handling text content and reasoning data.

johannes117 committed:
style: Refine MessageReasoning component for improved UI

- Removed unnecessary border styles and adjusted background colors for better visual consistency.
- Updated button styles for enhanced hover effects and improved text visibility.
- Simplified loading indicator styling to align with overall design updates.

johannes117 committed:
feat: Integrate Markdown rendering in MessageReasoning component

- Replaced plain text rendering of reasoning with a MarkdownRenderer for enhanced formatting and presentation.
- Updated the component to maintain visual consistency while improving the display of reasoning content.

johannes117 committed:
refactor: Simplify reasoning streaming logic in MessageReasoning component

- Removed the useEffect hook and related state management for streaming reasoning, streamlining the component's functionality.
- Cleaned up the code by eliminating unnecessary references, enhancing readability and maintainability.

johannes117 committed:
fix: Update Anthropic provider options to include budget tokens

- Added a budgetTokens property to the Anthropic provider options in the chat action, enhancing configuration for reasoning capabilities.

johannes117 committed:
feat: Add Switch component and enhance chat settings for reasoning capabilities

- Introduced a new Switch component for toggling reasoning features in the chat settings.
- Updated ChatRunSettings to include a toggle for enabling/disabling reasoning based on model capabilities.
- Enhanced ChatParams and related handlers to support the new isThinkingEnabled parameter for chat interactions.
- Modified the ModelConfig type to include canToggleThinking property, allowing for dynamic reasoning control based on model support.

johannes117 committed:
fix: Update BrainCircuit icon to Brain in ChatRunSettings component

- Replaced the BrainCircuit icon with the Brain icon for improved clarity in the reasoning toggle section of the chat settings.
- Ensured consistent icon usage throughout the component for better visual representation of features.

johannes117 committed:
style: Enhance layout and spacing in ChatRunSettings component

- Updated the layout of the ChatRunSettings component to improve visual structure and usability.
- Adjusted spacing and padding for various sections, including headers and content areas, to create a more organized appearance.
- Ensured consistent spacing for elements within the component, enhancing overall readability and user experience.

johannes117 committed:
style: Add no-scrollbar class to ChatRunSettings component for improved aesthetics

- Updated the ChatRunSettings component to include a no-scrollbar class in the scrollable content area, enhancing the visual appearance and user experience.

johannes117 committed:
style: Add no-scrollbar class to ChatSidebar component for improved aesthetics

- Updated the ChatSidebar component to include a no-scrollbar class in the scrollable content area, enhancing the visual appearance and user experience.

SamDebono commented on convex/ai.ts:
```diff
@@ -36,14 +36,18 @@ const messagePartValidator = v.union(
   _creationTime: v.number(),
   conversationId: v.id("conversations"),
-  role: v.union(v.literal(MESSAGE_ROLES.USER), v.literal(MESSAGE_ROLES.ASSISTANT), v.literal(MESSAGE_ROLES.SYSTEM), v.literal(MESSAGE_ROLES.DATA)),
+  role: v.union(v.literal("user" as MessageRole), v.literal("assistant" as MessageRole), v.literal("system" as MessageRole), v.literal("data" as MessageRole)),
```
why did you revert the enums?

SamDebono commented on convex/ai.ts:
```diff
@@ -274,29 +280,45 @@ export const chat = internalAction({
+        modelConfig.supportsReasoning &&
+        (isThinkingEnabled || !modelConfig.canToggleThinking)
+      ) {
+        if (provider === 'google') {
```
might want to move these to a constants and update the codebase I see it everywhere

SamDebono commented on convex/ai.ts:
```diff
@@ -274,29 +280,45 @@ export const chat = internalAction({
+      ) {
+        if (provider === 'google') {
+          providerOptions.google = { thinkingConfig: { includeThoughts: true } };
+        } else if (provider === 'anthropic') {
```
same as the above

SamDebono commented on frontend/components/Messages.tsx:
```diff
@@ -15,6 +15,21 @@ function PureMessages({
+  const lastMessage = messages.length > 0 ? messages[messages.length - 1] : null;
+  
+  // Determine if the assistant is generating a response
+  const assistantIsGenerating =
```
do we need to enter here if lastMessage is null? shouldnt we just jump to the return?

johannes117 committed:
refactor: Introduce PROVIDERS constant and update provider references throughout the codebase

- Added a new PROVIDERS constant in constants.ts to centralize provider string values.
- Updated references to provider strings in ai.ts, models.ts, and apiKeys.ts to use the new PROVIDERS constant for consistency and maintainability.
- Enhanced message role validation in ai.ts to utilize MESSAGE_ROLES for improved clarity.

johannes117 committed:
Merge branch 'main' into reasoning-ui
